src/can.rs

```diff
@@ -0,0 +1,272 @@
+    /// Transmit transaction
+    Transmit,
+    /// Receive transaction
+    Receive,
+}
+
+/// CAN Transaction type
+///
+/// Models an CAN transmit or receive transaction
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Transaction {
+    expected_mode: Mode,
+    expected_frame: Option<MockFrame>,
+    response_frame: Option<MockFrame>,
+    /// An optional error return for a transaction.
+    ///
+    /// This is in addition to the mode to allow validation that the
+    /// transaction mode is correct prior to returning the error.
+    expected_err: Option<ErrorKind>,
+}
+
+impl Transaction {
+    /// Create a Transmit transaction
+    pub fn transmit(expected_frame: MockFrame) -> Transaction {
+        Transaction {
+            expected_mode: Mode::Transmit,
+            expected_frame: Some(expected_frame),
+            response_frame: None,
+            expected_err: None,
+        }
+    }
+
+    /// Create a Receive transaction
+    pub fn receive(response_frame: MockFrame) -> Transaction {
+        Transaction {
+            expected_mode: Mode::Receive,
+            expected_frame: None,
+            response_frame: Some(response_frame),
+            expected_err: None,
+        }
+    }
+
+    /// Add an error return to a transaction
+    ///
+    /// This is used to mock failure behaviours.
+    pub fn with_error(mut self, error: ErrorKind) -> Self {
+        self.expected_err = Some(error);
+        self
+    }
+}
+
+/// Mock CAN Frame
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct MockFrame {
+    id: Id,
+    data: Vec<u8>,
+}
+
+impl Frame for MockFrame {
+    fn new(id: impl Into<Id>, data: &[u8]) -> Option<Self> {
+        let id = id.into();
+        let data = data.to_vec();
+        Some(Self { id, data })
+    }
+
+    fn new_remote(id: impl Into<Id>, _: usize) -> Option<Self> {
+        let id = id.into();
+        let data = [].to_vec();
+        Some(Self { id, data })
+    }
+
+    fn is_extended(&self) -> bool {
+        match self.id {
+            Id::Standard(_) => false,
+            Id::Extended(_) => true,
+        }
+    }
+
+    fn is_remote_frame(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    fn id(&self) -> Id {
+        self.id
+    }
+
+    fn dlc(&self) -> usize {
+        self.data.len()
+    }
+
+    fn data(&self) -> &[u8] {
+        &self.data
+    }
+}
+
+/// Mock CAN implementation
+pub type Mock = Generic<Transaction>;
+
+impl embedded_can::blocking::Can for Mock {
+    type Frame = MockFrame;
+    type Error = MockError;
+
+    fn transmit(&mut self, frame: &Self::Frame) -> Result<(), MockError> {
+        let t = self.next().expect("no expectation for can::transmit call");
+        assert_eq!(
+            t.expected_mode,
+            Mode::Transmit,
+            "can::transmit unexpected mode"
+        );
+        assert_eq!(
+            &t.expected_frame.unwrap(),
+            frame,
+            "can::transmit data does not match expectation"
+        );
```

Comment:
this looks pretty reasonable to me, tho i think is missing a path to return errors? (if `expected_error` is set this value should be returned instead of `Ok(())` (and the same for receive and the nb impls)